WebSockets Next: allow to select authentication mechanism and OIDC tenant used for opening WebSocket handshake with annotation (#46161)
quarkusio/quarkus: docs/src/main/asciidoc/websockets-next-reference.adoc

quarkus.http.auth.permission.http-upgrade.policy=authenticated
----

Security annotations used during authentication must be placed on an endpoint class as well, for the `SecurityIdentity` is created before the websocket connection is opened.

.Select Bearer token authentication mechanism
[source, java]
----
package io.quarkus.websockets.next.test.security;

import io.quarkus.oidc.BearerTokenAuthentication;
import io.quarkus.websockets.next.OnTextMessage;
import io.quarkus.websockets.next.WebSocket;

@BearerTokenAuthentication <1>
@WebSocket(path = "/end")
public class Endpoint {

    @OnTextMessage
    String echo(String message) {
        return message;
    }

}
----

> <1> Require that opening WebSocket handshake request is authenticated using Bearer token authentication.

Reviewer: ```suggestion
<1> Require that an opening WebSocket handshake request is authenticated using the bearer token authentication.
```

As far as I recall, the Doc team prefers a lower case when referring to it in the text

> Find how to select other authentication mechanisms with annotations in the xref:security-authentication-mechanisms.adoc#use-annotations-for-path-based-auth[Authentication mechanisms in Quarkus] guide.

Reviewer: ```suggestion
See the xref:security-authentication-mechanisms.adoc#use-annotations-for-path-based-auth[Authentication mechanisms in Quarkus] guide for more information about selecting authentication mechanisms with annotations.
```

[source,properties]
----
quarkus.http.auth.proactive=false <1>
----

> <1> Ensures that Quarkus matches opening WebSocket handshake request with the `io.quarkus.oidc.BearerTokenAuthentication` annotation before authentication begins.

Reviewer: ```suggestion
<1> Start authenticating an opening WebSocket handshake request only when the `io.quarkus.oidc.BearerTokenAuthentication` annotation is detected.
```

> Another example of a security annotation that must be placed on the class-level is the `io.quarkus.oidc.Tenant` annotation, which is used for resolving the OIDC tenant identifiers.

Reviewer: Michal, if we do talk about `@Tenant` here, than your original code example showing it would've been best, but you agreed to remove it from the actual code. I'm sorry for being pedantic :-), but the OIDC multi-tenancy is not relevant in this doc, OIDC users who work with the multi-tenancy will find your update in the OIDC multi-tenancy doc of help. 
(I think the situation where we have WebSocket Next combined with more than one OIDC provider a very advanced and therefore rather rare case).
Proposing to drop this block

Author: np, let's drop it

quarkus.http.auth.permission.http-upgrade.policy=authenticated
----

Security annotations used during authentication must be placed on an endpoint class as well, for the `SecurityIdentity` is created before the websocket connection is opened.

.Select Bearer token authentication mechanism and OIDC tenant with annotation
[source, java]
----
package io.quarkus.websockets.next.test.security;

import io.quarkus.oidc.BearerTokenAuthentication;
import io.quarkus.oidc.Tenant;
import io.quarkus.websockets.next.OnTextMessage;
import io.quarkus.websockets.next.WebSocket;

@BearerTokenAuthentication <1>

> @Tenant("hr") <2>

Reviewer: I wonder if it is a bit too OIDC specific for the WebSockets Next guide ? The PR code updates related to it may be fine but I'm not sure we should go into OIDC multi-tenancy is this doc

Author: We must make it clear that these annotations should only be used on HTTP upgrade and not on methods like `@OnTextMessage void onTextMessage()` (because it's too late). There is build-time validation that prevents users to use it on methods and also, if this validation somehow failed, there is runtime validation that prevents this as well (so literally impossible for it to happen). but still, I think it is WS Next specific, this is only extension where we use class-level security so far. I don't want to confuse users?

Author: to be clear, I am open to any changes, I just think it is very important information, I want users to understand HOW and WHY (both is there - there is text that explain when auth happens)

Reviewer: Sure, it is just that is a very OIDC specific content, `BearerTokenAuthentication` can be supported by smallrye-jwt, may be by custom mechanisms, but `@Tenant` is a multi-tenant OIDC specific detail.

Perhaps, you should add a note in the OIDC tenancy doc instead, in the section which describes `@Tenant`, and link to the section you are adding here ?

Author: > Sure, it is just that is a very OIDC specific content, `BearerTokenAuthentication` can be supported by smallrye-jwt, may be by custom mechanisms, but `@Tenant` is a multi-tenant OIDC specific detail.

true

> Perhaps, you should add a note in the OIDC tenancy doc instead, in the section which describes `@Tenant`, and link to the section you are adding here ?

I'll do that.

Author: Done, I'll resolve this as we can address current state elsewhere. Thanks for the suggestion.
